lgc/patch/PatchEntryPointMutate.cpp

```diff
@@ -630,7 +672,113 @@ void PatchEntryPointMutate::processShader(ShaderInputs *shaderInputs) {
+// Process all functions in a compute pipeline or library.
+//
+// @param shaderInputs : ShaderInputs object representing hardware-provided shader inputs
+// @param [in/out] module : Module
+void PatchEntryPointMutate::processComputeFuncs(ShaderInputs *shaderInputs, Module &module) {
+  m_shaderStage = ShaderStageCompute;
+
+  // Determine what args need to be added on to all functions.
+  SmallVector<Type *, 8> argTys;
+  uint64_t inRegMask = generateEntryPointArgTys(shaderInputs, argTys);
+
+  // Process each function definition.
+  SmallVector<Function *, 4> origFuncs;
+  for (Function &func : module) {
+    if (!func.isDeclaration())
+      origFuncs.push_back(&func);
+  }
+  for (Function *origFunc : origFuncs) {
+    // Create the new function and transfer code and attributes to it.
+    Function *newFunc = addFunctionArgs(origFunc, origFunc->getFunctionType()->getReturnType(), argTys, inRegMask);
+
+    // Set Attributes on new function.
+    setFuncAttrs(newFunc);
+
+    // Change any uses of the old function to a bitcast of the new function.
+    SmallVector<Use *, 4> funcUses;
+    for (auto &use : origFunc->uses())
+      funcUses.push_back(&use);
+    Constant *bitCastFunc = ConstantExpr::getBitCast(newFunc, origFunc->getType());
+    for (Use *use : funcUses)
+      *use = bitCastFunc;
+
+    // Remove original function.
+    origFunc->eraseFromParent();
+  }
+
+  if (origFuncs.size() == 1 && !isComputeWithCalls())
+    return;
+
+  // This is one of:
+  // - a compute pipeline with non-inlined functions;
+  // - a compute pipeline with calls to library functions;
+  // - a compute library.
+  // We need to scan the code and modify each call to prepend the extra args.
+  IRBuilder<> builder(module.getContext());
+  for (Function &func : module) {
+    if (func.isDeclaration())
+      continue;
+    for (BasicBlock &block : func) {
+      for (auto it = block.begin(), end = block.end(); it != end;) {
+        // Get the instruction and increment the iterator beyond it, so we can safely erase the instruction.
+        Instruction *inst = &*it;
+        ++it;
+        auto call = dyn_cast<CallInst>(inst);
+        if (!call)
+          continue;
+        // Got a call. Skip it if it calls an intrinsic or an internal lgc.* function.
+        Value *calledVal = call->getCalledOperand();
+        if (auto calledFunc = dyn_cast<Function>(calledVal)) {
+          if (calledFunc->isIntrinsic() || calledFunc->getName().startswith(lgcName::CallPrefix))
+            continue;
+        }
+        // Build a new arg list, made of the ABI args shared by all functions (user data and hardware shader
+        // inputs), plus the original args on the call.
+        SmallVector<Type *, 8> argTys;
+        SmallVector<Value *, 8> args;
+        Function *func = call->getFunction();
+        for (unsigned idx = 0; idx != argTys.size(); ++idx) {
```

Comment:
Is this meant to be `func->arg_size()`? `argTys` is still empty here.
Also, is `func`, the callee which is assigned the line above, the right thing in the for loop?

I think something like `for (Function &caller : module)` and using `Function *callee = call->getFunction()` would be more obvious.

Reply:
Oops, I confused myself with two instances of creating a second variable with the same name:

1. The argTys I want to use in the condition in the for is meant to be the one at the top of the function, the types of the args passed into the shader entry-point, not the one I defined 3 lines higher. I will rename the first one "shaderInputTys".

2. call->getFunction() gets the function the call (or any instruction) is enclosed in, not the function it is calling. So this "func" is exactly the same as the other "func" higher up, except the first one is a ref and the second one is a pointer. I will remove the second one.